File: tools/ceremony/key.go
package main

import (
	"crypto/sha256"
	"encoding/base32"
	"encoding/base64"
	"errors"
	"fmt"
	"runtime"
	"strings"
	"time"

	"github.com/cosmos/cosmos-sdk/crypto/hd"
	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
	"github.com/cosmos/cosmos-sdk/types/bech32"
	bip39 "github.com/cosmos/go-bip39"
)

// secret is a recovery phrase held in a buffer this program owns.
//
// The claim this type supports is narrow and worth stating precisely, because
// the runbook depends on it and the report says the same thing:
//
// What it guarantees. Nothing in this program writes the phrase to a file, a
// log, an environment variable or a command line, and it is never a function
// argument outside this file. The only sink is console.print, which refuses to
// run when its output is not a terminal.
//
// What it cannot guarantee. bip39.NewMnemonic returns a Go string, and a Go
// string is immutable: that copy stays wherever the allocator put it until the
// garbage collector reuses the page, and no code in Go can shorten that. The
// same is true of the string handed to hd.Secp256k1.Derive, which is the
// chain's own derivation path and the reason this tool does not reimplement it
// — a wallet that derived addresses even slightly differently would produce
// keys that look correct and control nothing. So: every buffer this program
// allocates is zeroed, and two short-lived immutable copies are not. The
// mitigations for those are a machine with no swap file, no hibernation and no
// crash dumps, and a power cycle at the end — which is why the runbook makes
// all four ceremony steps rather than recommendations.
type secret struct {
	buf []byte
}

// newSecret generates a fresh phrase.
func newSecret() (*secret, error) {
	entropy, err := bip39.NewEntropy(entropyBits)
	if err != nil {
		return nil, err
	}
	defer zero(entropy)

	phrase, err := bip39.NewMnemonic(entropy)
	if err != nil {
		return nil, err
	}
	// Copied into a buffer immediately so that everything downstream of this
	// line is zeroable. See the type comment for what this does not fix.
	return &secret{buf: []byte(phrase)}, nil
}

// secretFromInput takes a phrase somebody has typed back in, for the restore
// drill.
func secretFromInput(line string) (*secret, error) {
	buf := []byte(strings.Join(strings.Fields(line), " "))

	// MnemonicToByteArray, not IsMnemonicValid.
	//
	// This is not a stylistic choice and the difference is the whole point of
	// the drill. bip39.IsMnemonicValid — the obvious-looking call, and the one
	// tools/wallet used — checks only that there are 12/15/18/21/24 words and
	// that each one appears in the wordlist. It does not verify the checksum.
	// Every realistic transcription error passes it: somebody copying by hand
	// writes another real BIP-39 word, so the word count is right, every word
	// is in the list, and the phrase validates while deriving a completely
	// different, empty account.
	//
	// MnemonicToByteArray recomputes the checksum, which is the four bits at
	// the end of a 24-word phrase whose entire job is catching exactly this.
	entropy, err := bip39.MnemonicToByteArray(string(buf))
	if err != nil {
		zero(buf)
		return nil, errors.New(
			"that is not a valid recovery phrase — the checksum does not match.\n" +
				"One of the words is wrong, or two are swapped. Check it against the sheet word\n" +
				"by word. Do not correct a guess and try again: a phrase with one wrong word\n" +
				"derives a different, empty account rather than failing, which is why the last\n" +
				"word of a BIP-39 phrase carries a checksum and why this is checked here")
	}
	zero(entropy)
	return &secret{buf: buf}, nil
}

// zero overwrites the phrase. Call it with defer, at the point the secret is
// created, every time.
func (s *secret) zero() {
	if s == nil {
		return
	}
	zero(s.buf)
	s.buf = nil
}

// wordCount is how many words the phrase has.
func (s *secret) wordCount() int {
	return len(splitWords(s.buf))
}

// word returns the i-th word, one-based, as a slice into the phrase buffer.
//
// A slice rather than a string so that displaying and checking the phrase
// allocates no copies of it that zero() would then miss.
func (s *secret) word(i int) []byte {
	words := splitWords(s.buf)
	if i < 1 || i > len(words) {
		return nil
	}
	return words[i-1]
}

func splitWords(buf []byte) [][]byte {
	var words [][]byte
	start := -1
	for i := 0; i <= len(buf); i++ {
		if i == len(buf) || buf[i] == ' ' {
			if start >= 0 {
				words = append(words, buf[start:i])
				start = -1
			}
			continue
		}
		if start < 0 {
			start = i
		}
	}
	return words
}

// zero overwrites a buffer and keeps it alive across the loop.
//
// runtime.KeepAlive is not decoration. Without it the compiler is free to
// notice that nothing reads the buffer afterwards and drop the stores, which
// would leave the phrase in memory while this function claims to have removed
// it. Go has no explicit_bzero; this is the closest thing available.
func zero(b []byte) {
	for i := range b {
		b[i] = 0
	}
	runtime.KeepAlive(b)
}

// derive turns the phrase into a key, through the chain's own derivation path.
//
// The caller owns the returned private key and must zero it. The one string
// conversion here is the unavoidable copy described on the secret type.
func (s *secret) derive(index uint32) (*secp256k1.PrivKey, string, error) {
	path := hd.NewFundraiserParams(0, coinType, index).String()

	derived, err := hd.Secp256k1.Derive()(string(s.buf), "", path)
	if err != nil {
		return nil, "", err
	}
	defer zero(derived)

	priv, ok := hd.Secp256k1.Generate()(derived).(*secp256k1.PrivKey)
	if !ok {
		return nil, "", errors.New("the chain's key generator returned an unexpected key type")
	}
	return priv, path, nil
}

// role is what a key is being generated for. It changes the warnings, the
// derived addresses shown, and what the ceremony record calls the holder.
type role string

const (
	roleCustodian role = "custodian"
	roleValidator role = "validator-operator"
)

// pubKeyJSON is the pubkey in the form every SDK message and query uses, so it
// can be pasted into a validator.json or a group members file without being
// re-encoded by hand.
type pubKeyJSON struct {
	Type string `json:"@type"`
	Key  string `json:"key"`
}

// identity is everything this ceremony publishes about one key. All of it is
// public: an address, a pubkey and a digest of the pubkey. Nothing here helps
// anybody spend anything.
type identity struct {
	Name        string     `json:"name"`
	Role        role       `json:"role"`
	Address     string     `json:"address"`
	Valoper     string     `json:"valoper,omitempty"`
	PubKey      pubKeyJSON `json:"pubkey"`
	Fingerprint string     `json:"fingerprint"`
	HDPath      string     `json:"hd_path"`
	GeneratedAt string     `json:"generated_at"`
	Ceremony    string     `json:"ceremony,omitempty"`
}

// identityOf assembles the public record for a derived key.
func identityOf(name string, r role, priv *secp256k1.PrivKey, path string, now time.Time) (identity, error) {
	pub := priv.PubKey()
	addressBytes := pub.Address().Bytes()

	address, err := bech32.ConvertAndEncode(accountPrefix, addressBytes)
	if err != nil {
		return identity{}, err
	}

	id := identity{
		Name: name,
		Role: r,
		PubKey: pubKeyJSON{
			Type: "/cosmos.crypto.secp256k1.PubKey",
			Key:  base64.StdEncoding.EncodeToString(pub.Bytes()),
		},
		Address:     address,
		Fingerprint: fingerprint(pub.Bytes()),
		HDPath:      path,
		GeneratedAt: now.UTC().Format(time.RFC3339),
	}

	// A validator's operator address is the same key in the valoper prefix.
	// Shown only for that role, because a custodian who was given one would
	// reasonably conclude the foundation key is also a staking key.
	if r == roleValidator {
		id.Valoper, err = bech32.ConvertAndEncode(accountPrefix+"valoper", addressBytes)
		if err != nil {
			return identity{}, err
		}
	}
	return id, nil
}

// fingerprintDomain separates this digest from every other digest of the same
// pubkey. Without it, a value printed here could be mistaken for — or replayed
// as — an address hash or a transaction hash somewhere else.
const fingerprintDomain = "yamale-ceremony-fingerprint-v1"

// fingerprint is the short string a custodian writes on their own sheet.
//
// It is what makes a swapped or mis-filed envelope detectable. Five years from
// now, an envelope labelled "custodian 3" either recovers to a key whose
// fingerprint matches the one on the ceremony record or it does not, and that
// check needs no network, no node and nothing anybody has to be trusted about.
//
// Crockford's base32 alphabet, so the four characters people confuse on paper —
// I, L, O, U — never appear, and 1/l, 0/O cannot be misread into a different
// valid value. Forty bits in two groups of four: short enough to be read aloud
// across a room and compared by eye, long enough that two of the five
// custodians colliding is not a thing that happens.
func fingerprint(pubkey []byte) string {
	sum := sha256.Sum256(append([]byte(fingerprintDomain), pubkey...))
	encoder := base32.NewEncoding("0123456789ABCDEFGHJKMNPQRSTVWXYZ").WithPadding(base32.NoPadding)
	encoded := encoder.EncodeToString(sum[:5])
	return encoded[:4] + "-" + encoded[4:8]
}

// describe is the human-readable line the record and the console both use.
func (id identity) describe() string {
	return fmt.Sprintf("%s  %s  %s", id.Fingerprint, id.Address, id.Name)
}

File: clients/ceremony/src/key.ts
// One custodian's key, generated in their own browser and never leaving it.
//
// This is the file the whole hosted ceremony rests on. A coordinator serves the
// page; the page generates the phrase here, derives the key here, signs here,
// and posts back a public key and two signatures. Nothing in this module has a
// path to the network, and nothing anywhere in this bundle sends a phrase or a
// private key: see src/storage.test.ts, which asserts that no request body the
// client builds contains either.
//
// The derivation has to match tools/ceremony/key.go exactly. A page deriving
// even slightly differently would hand five custodians addresses that look right
// and control nothing, and the failure would not surface until somebody tried to
// spend from the account that holds every seized asset on the chain. That is why
// nothing here is written from first principles: BIP-39, BIP-32 and secp256k1
// are @scure and @noble, and the outputs are pinned against the Go binary by
// testdata/vectors/ceremony.json.

import { HDKey } from '@scure/bip32';
import { generateMnemonic, mnemonicToSeedSync, validateMnemonic } from '@scure/bip39';
import { wordlist } from '@scure/bip39/wordlists/english';
import { secp256k1 } from '@noble/curves/secp256k1.js';
import { sha256 } from '@noble/hashes/sha2.js';
import { ripemd160 } from '@noble/hashes/legacy.js';
import { bech32 } from '@scure/base';
import {
  FINGERPRINT_DOMAIN,
  SECP256K1_PUBKEY_TYPE,
  possessionMessage,
  shortDigest,
  toBase64,
  type Identity,
  type Submission,
} from './wire.ts';

// ACCOUNT_PREFIX and COIN_TYPE are what make a key a Yamale key. Restated here
// rather than fetched from the coordinator: a page that asked the server which
// prefix to use would let a server that answered "cosmos" hand five custodians
// keys for a different chain, and every check afterwards would agree because
// they would all read the same answer.
export const ACCOUNT_PREFIX = 'yml';
export const COIN_TYPE = 118;

// ENTROPY_BITS is 256, so twenty-four words. Twelve is legal and this page does
// not offer it: the extra sixty-four bits cost one line on a sheet of paper that
// will be written once and read almost never, and the choice is not one a
// custodian should be reasoning about while four other people wait.
export const ENTROPY_BITS = 256;

// BECH32_LIMIT is above the default 90 because a group policy address is
// thirty-two bytes and encodes to sixty-two characters. The default would refuse
// the one address this ceremony exists to produce.
const BECH32_LIMIT = 256;

export function hdPath(index: number): string {
  return `m/44'/${COIN_TYPE}'/0'/0/${index}`;
}

export function newPhrase(): string {
  return generateMnemonic(wordlist, ENTROPY_BITS);
}

// checkPhrase validates the checksum, not just the words.
//
// This is the difference the restore drill turns on. A check that only confirmed
// each word is in the wordlist would pass every realistic transcription error:
// somebody copying by hand writes another real BIP-39 word, so the count is
// right, every word is in the list, and the phrase derives a completely
// different, empty account. The last four bits of a 24-word phrase exist to
// catch exactly that.
export function checkPhrase(phrase: string): boolean {
  return validateMnemonic(normalisePhrase(phrase), wordlist);
}

export function normalisePhrase(phrase: string): string {
  return phrase.trim().toLowerCase().split(/\s+/).filter(Boolean).join(' ');
}

export type KeyPair = {
  priv: Uint8Array;
  pub: Uint8Array;
  address: string;
  fingerprint: string;
  path: string;
};

export function deriveKey(phrase: string, index: number): KeyPair {
  const path = hdPath(index);
  // No BIP-39 passphrase. The chain's own derivation uses an empty one, and a
  // ceremony that offered a passphrase would be a ceremony where a custodian
  // holds a sheet that does not, on its own, recover the key.
  const seed = mnemonicToSeedSync(normalisePhrase(phrase), '');
  const node = HDKey.fromMasterSeed(seed).derive(path);
  if (!node.privateKey || !node.publicKey) {
    throw new Error('the derivation produced no key');
  }
  const priv = node.privateKey;
  const pub = node.publicKey;
  return {
    priv,
    pub,
    address: bech32Address(ACCOUNT_PREFIX, addressBytes(pub)),
    fingerprint: fingerprintOf(pub),
    path,
  };
}

// addressBytes is the Cosmos SDK's secp256k1 account address: RIPEMD-160 of
// SHA-256 of the compressed public key.
export function addressBytes(pub: Uint8Array): Uint8Array {
  return ripemd160(sha256(pub));
}

export function bech32Address(prefix: string, raw: Uint8Array): string {
  return bech32.encode(prefix, bech32.toWords(raw), BECH32_LIMIT);
}

export function decodeBech32(address: string): { prefix: string; bytes: Uint8Array } {
  const decoded = bech32.decode(address as `${string}1${string}`, BECH32_LIMIT);
  return { prefix: decoded.prefix, bytes: Uint8Array.from(bech32.fromWords(decoded.words)) };
}

// fingerprintOf is the short string a custodian writes on their own sheet.
//
// It is what makes a swapped or mis-filed envelope detectable years later: an
// envelope labelled "custodian 3" either recovers to a key whose fingerprint
// matches the one on the record or it does not, and that check needs no network
// and nothing anybody has to be trusted about.
export function fingerprintOf(pub: Uint8Array): string {
  return shortDigest(FINGERPRINT_DOMAIN, pub);
}

export function identityOf(name: string, key: KeyPair, generatedAt: Date): Identity {
  return {
    name,
    role: 'custodian',
    address: key.address,
    pubkey: { '@type': SECP256K1_PUBKEY_TYPE, key: toBase64(key.pub) },
    fingerprint: key.fingerprint,
    hd_path: key.path,
    generated_at: rfc3339Second(generatedAt),
  };
}

// rfc3339Second is the timestamp format the Go side writes and signs.
//
// Truncated to the second, in UTC, with a Z rather than an offset. The latest of
// the five generated_at values ends up inside the genesis fragment, which the
// group fingerprint covers, so a browser emitting milliseconds would give every
// custodian a different fingerprint from the same five submissions.
export function rfc3339Second(when: Date): string {
  return `${when.toISOString().slice(0, 19)}Z`;
}

// sign produces a Cosmos secp256k1 signature: 64 bytes of r||s, low-S
// normalised, over the SHA-256 of the message.
//
// Low-S matters because the chain's verifier rejects the high-S form of the same
// signature. A page that emitted it would produce submissions that verified
// nowhere, roughly half the time, which is the worst possible failure rate: too
// often to be a fluke, rarely enough to look like a network problem.
export function sign(message: Uint8Array, priv: Uint8Array): Uint8Array {
  return secp256k1.sign(sha256(message), priv, { lowS: true }).toCompactRawBytes();
}

export function verify(signature: Uint8Array, message: Uint8Array, pub: Uint8Array): boolean {
  try {
    return secp256k1.verify(signature, sha256(message), pub, { lowS: true });
  } catch {
    return false;
  }
}

export function signSubmission(ceremonyID: string, id: Identity, priv: Uint8Array): Submission {
  return {
    ceremony_id: ceremonyID,
    identity: id,
    possession: toBase64(sign(possessionMessage(ceremonyID, id), priv)),
  };
}

// zero overwrites a key buffer.
//
// Worth stating what this does and does not achieve, because the page says so
// too. It clears the one copy this code holds. It cannot clear a copy the
// JavaScript engine made — strings are immutable and the garbage collector
// decides when a page is reused — and no code in a browser can. The mitigation
// for that is the browser being closed, which is why the last screen says to
// close the tab rather than implying the page has cleaned up after itself.
export function zero(buf: Uint8Array): void {
  buf.fill(0);
}
